### DISCUSSION TITLE
VGAE or GAE with weighted matrixes

### DISCUSSION THREAD

**Discussion by TommyGiak** [pyg-team/pytorch_geometric#7631] [Ideas]
I was searching how to implement a VGAE using weighted matrixes. I found that this is implemented only for the encoder (I used a GCNconv), but after the decoder of the VGAE there is no implementation for weights in the recon_loss, is it right?\
Do I need to implement it by myself?\
Anyway I will try to implement it if it's possible.

Thanks for the help!

**Comment by rusty1s** [comment]
Yes, the decoder in VGAE has no learnable parameters (it is just an inner product). You can make this decoder learnable, e.g., via
```python
MLP(x_src, x_dst) -> [0, 1]
```
but using an inner product decoder is the most common decoder.

**Reply by TommyGiak** [reply]
Actually, I mean another thing, my question is not clear.
What I want is to build a VGAE for a weighted matrix, for instance a weighted matrix like this:
```python
link_index = torch.tensor([[0, 1, 0, 3, 4], 
                           [1, 2, 3, 1, 2]])
edge_weight = torch.tensor([0.2, 1, 0.6, 0.8, 2])
```
When I create the encoder (with GCNconv) there's no problem since the forward call I can include the edge_weight tensor:
```python
def forward(self, x: Tensor, edge_index: Adj,
            edge_weight: OptTensor = None) -> Tensor:
```
Then I can build the VGAE using the correct GCNEncoder class. If now I call a forward step of the VGAE I can include also the edge_weight since the VGAE has:
```python
forward(*args, **kwargs) -> Tensor    #alias for encode 
```
So up to now I can encode and then decode a matrix with also the information of the edge_index tensor.\
The problem comes when I want to train the VGAE: the recon_loss of the VGAE require only pos_edge_index (and optionally neg_edge_index) and it doesn't require the edge_weight:
```python
def recon_loss(self, z: Tensor, pos_edge_index: Tensor,
               neg_edge_index: Optional[Tensor] = None) -> Tensor:
```
so the VGAE during the training tries to rebuilt the adj matrix only with ones and zeros for pos_index and neg_index and it doesn't try to rebuild the adj matrix with the correct edge_weight.\
So what I'm asking is if there exist an implementation of a VGAE for the weighted adj matrixes.\
I hope now is more clear :sweat_smile:
Thanks

**Reply by rusty1s** [reply]
I think you can implement your own `recon_loss` function that is able to achieve this. You can still use binary cross entropy for training in case edge weights are in between [0, 1]. Note that you also need to create an edge weight for negative edges (e.g., via sampling from `torch.rand`).
